File: java/linear_data_structures/src/main/java/com/qe/test/ds/CircularQueue.java
package com.qe.test.ds;

import java.util.NoSuchElementException;

public class CircularQueue<E> {
    private final Object[] buffer;
    private int head;
    private int tail;
    private int count;

    @SuppressWarnings("unchecked")
    public CircularQueue(int capacity) {
        if (capacity <= 0) throw new IllegalArgumentException("capacity must be positive");
        this.buffer = new Object[capacity];
        this.head = 0;
        this.tail = 0;
        this.count = 0;
    }

    public void enqueue(E item) {
        if (item == null) throw new IllegalArgumentException("item must not be null");
        if (isFull()) throw new IllegalStateException("queue is full");
        buffer[tail] = item;
        tail = (tail + 1) % buffer.length;
        count++;
    }

    @SuppressWarnings("unchecked")
    public E dequeue() {
        if (isEmpty()) throw new NoSuchElementException("queue is empty");
        var item = (E) buffer[head];
        buffer[head] = null;
        head = (head + 1) % buffer.length;
        count--;
        return item;
    }

    @SuppressWarnings("unchecked")
    public E peek() {
        if (isEmpty()) throw new NoSuchElementException("queue is empty");
        return (E) buffer[head];
    }

    public boolean isEmpty() {
        return count == 0;
    }

    public boolean isFull() {
        return count == buffer.length;
    }

    public int size() {
        return count;
    }

    public int capacity() {
        return buffer.length;
    }
}

File: python/linear_data_structures/implementation/circular_queue.py
from __future__ import annotations

from typing import Generic, TypeVar

T = TypeVar("T")


class CircularQueue(Generic[T]):
    def __init__(self, capacity: int) -> None:
        if capacity < 1:
            raise ValueError("capacity must be >= 1")
        self._buffer: list[T | None] = [None] * capacity
        self._capacity = capacity
        self._head = 0
        self._tail = 0
        self._count = 0

    def enqueue(self, item: T) -> None:
        if self.is_full:
            raise IndexError("enqueue on full circular queue")
        self._buffer[self._tail] = item
        self._tail = (self._tail + 1) % self._capacity
        self._count += 1

    def dequeue(self) -> T:
        if self.is_empty:
            raise IndexError("dequeue from empty circular queue")
        value = self._buffer[self._head]
        self._buffer[self._head] = None
        self._head = (self._head + 1) % self._capacity
        self._count -= 1
        return value  # type: ignore[return-value]

    def peek(self) -> T:
        if self.is_empty:
            raise IndexError("peek from empty circular queue")
        return self._buffer[self._head]  # type: ignore[return-value]

    @property
    def is_full(self) -> bool:
        return self._count == self._capacity

    @property
    def is_empty(self) -> bool:
        return self._count == 0

    @property
    def size(self) -> int:
        return self._count

    @property
    def capacity(self) -> int:
        return self._capacity
